### DISCUSSION TITLE
Data source type

### DISCUSSION THREAD

**Discussion by aiamk** [Cecilapp/Cecil#1672] [Q&A] [answered]
Hello, 

Cecila looks promising, but is it possible to extend it to have different sources type rather than files, i mean data source either from database or rest api?

thank you..

best regards

**Comment by ArnaudLigny** [comment]
Hello @aiamk,

Cecil creates its pages from Markdown files from the `pages` directory, and also creates "virtual" pages (like sections, pagination, etc.) with generators (see the following the life cycle schema: https://github.com/Cecilapp/Cecil/blob/master/docs/charts/cecil-build-life-cycle.png).

So you can create your own generator to create new pages (or alter alrady created pages) from diffents sources like an API or a database.
This is not really obvious, but possible: https://cecil.app/documentation/configuration/#custom-generator

I hope it help.

**Comment by aiamk** [comment]
i saw this page previously, but its not clear and example for single page, i am looking to read the collection from database.

do i have to implement the data collection from database within the extend custom generator class and iterate it?

thank you

**Reply by ArnaudLigny** [reply, marked as answer]
Yes, if you needs to create a collection of pages you can do it from a custom generator.

SQLite example:

```php
<?php
namespace Cecil\Generator;

use Cecil\Collection\Page\Page;
use Cecil\Collection\Page\Type;

class MyGenerator extends AbstractGenerator implements GeneratorInterface
{
    public function generate(): void
    {
        $db = new SQLite3('database.sqlite');
		$statement = $db->prepare('SELECT ...');
		$result = $statement->execute();

		while ($row = $result->fetchArray(SQLITE3_ASSOC)) {
			$page = (new Page($row['page-id']))
	        	->setType(Type::PAGE)
	          	->setPath($row['path'])
	          	->setBodyHtml($row['html'])
	          	->setVariable('title', $row['title'])
	          	->setVariable('date', $row['date'])
			$this->generatedPages->add($page);
		}

        $result->finalize();
		$db->close();
    }
}
```

**Comment by aiamk** [comment]
Thank you very much @ArnaudLigny , i appreciate  your help.

Best Regards
